serve/graph/model/transaction.go

```diff
@@ -2,43 +2,228 @@ package model
 
 import (
 	"encoding/base64"
+	"encoding/json"
 	"fmt"
 
+	"github.com/gnolang/gno/gno.land/pkg/sdk/vm"
+	"github.com/gnolang/gno/tm2/pkg/amino"
 	"github.com/gnolang/gno/tm2/pkg/bft/types"
+	"github.com/gnolang/gno/tm2/pkg/sdk/bank"
+	"github.com/gnolang/gno/tm2/pkg/std"
 )
 
 type Transaction struct {
-	t *types.TxResult
+	memo     string
+	txResult *types.TxResult
+	messages []*TransactionMessage
 }
 
-func NewTransaction(t *types.TxResult) *Transaction {
-	return &Transaction{t: t}
+func NewTransaction(txResult *types.TxResult) *Transaction {
+	var stdTx std.Tx
+
+	if err := amino.Unmarshal(txResult.Tx, &stdTx); err != nil {
+		return nil
+	}
+
+	messages := make([]*TransactionMessage, 0)
+	for _, message := range stdTx.GetMsgs() {
+		messages = append(messages, NewTransactionMessage(message))
+	}
+
+	return &Transaction{txResult: txResult, messages: messages, memo: stdTx.GetMemo()}
 }
 
 func (t *Transaction) ID() string {
-	return fmt.Sprintf("%d_%d", t.t.Height, t.t.Index)
+	return fmt.Sprintf("%d_%d", t.txResult.Height, t.txResult.Index)
 }
 
 func (t *Transaction) Index() int {
-	return int(t.t.Index)
+	return int(t.txResult.Index)
 }
 
 func (t *Transaction) Hash() string {
-	return base64.StdEncoding.EncodeToString(t.t.Tx.Hash())
+	return base64.StdEncoding.EncodeToString(t.txResult.Tx.Hash())
 }
 
 func (t *Transaction) BlockHeight() int {
-	return int(t.t.Height)
+	return int(t.txResult.Height)
 }
 
 func (t *Transaction) GasWanted() int {
-	return int(t.t.Response.GasWanted)
+	return int(t.txResult.Response.GasWanted)
 }
 
 func (t *Transaction) GasUsed() int {
-	return int(t.t.Response.GasUsed)
+	return int(t.txResult.Response.GasUsed)
 }
 
 func (t *Transaction) ContentRaw() string {
-	return t.t.Tx.String()
+	return t.txResult.Tx.String()
+}
+
+func (t *Transaction) Memo() string {
+	return t.memo
+}
+
+func (t *Transaction) Messages() []*TransactionMessage {
+	return t.messages
+}
+
+func (t *Transaction) Fee() *TxFee {
+	return &TxFee{
+		GasWanted: t.GasWanted(),
+		GasFee:    t.GasUsed(),
+	}
+}
+
+type TransactionMessage struct {
+	Value   MessageValue
+	TypeURL MessageType
+	Route   MessageRoute
+}
+
+func NewTransactionMessage(message std.Msg) *TransactionMessage {
+	var contentMessage *TransactionMessage
+
+	switch message.Route() {
+	case bank.RouterKey:
+		if message.Type() == MessageTypeSend.String() {
+			contentMessage = &TransactionMessage{
+				Route:   MessageRouteBank,
+				TypeURL: MessageTypeSend,
+				Value:   makeBankMsgSend(message),
+			}
+		}
```

Comment:
What if the message type is not TypeSend? Do we lose the message?

Maybe we need a message type unknown for these cases, just in case we receive something that the indexer does not understand yet, to avoid the case that we are hiding information to the user.

The unknown message will only contain the type, route and the rawContent.

Reply:
I changed it to show the type and raw data if the message was not generated.